SalesforceCommerceCloud/pwa-kit
# [Feature] Shopper Experience / Page Designer Components

This PR contains changes from #958. Be aware to when reviewing so you aren't commenting on those changes.

# Description

This PR add the basic building block components for Page Designer into the `commerce-sdk-react` library. The components added are listed below. 

- **Page**: The top level component that maps to the `page` asset made available via the Shopper Experience `getPage` and `getPages` endpoints. The purpose of this component is to allow developer to render their Page Designer Pages with one single component.
- **Regions**: This is also a top level component, but isn't directly mapped to a top level data structure from the API. This container component is used internally in the Page component above to render the components of a given Page Designer page object. It's also important to note that this component can and should be used when creating new components that represent page designer components objects. For example, if you are creating a component that simple adds chrome around it's children components, you can use the Region component directly to simplify your components implementation.
- **Components**: This component really isn't much of a component, but instead is used to locate a component within the pages components map and render it with the correct data. _NOTE: This component is only used internally._

**The reason I chose to place these components in the `commerce-sdk-react` is because these aren't components that are ever going to be overwritten or customized by the end user and it also allows us to have control of upgrading if there are bugs are feature requests. **

## Example Usage:

Given the following page object as returned from the `getPage` endpoint, this is how you might use that component.

```
[{
    "id": "homepage",
    "typeId": "storePage",
    "aspectTypeId": "pdpAspect",
    "name": "Homepage",
    "description": "Home page of the storefront.",
    "pageTitle": "Homepage - My Shop",
    "pageDescription": "Get started shopping",
    "pageKeywords": "fashion, my shop",
    "regions": [{
            "id": "header",
            "components": [{
                "id": "iofwj38fhw3f",
                "typeId": "commerce_assets.banner",
                "data": {
                    "title": "Products On Sale",
                    "bannerImage": "sale/topsellerPromo.jpg"
                }
            }]
        },
        {
            "id": "main",
            "components": [{
                "id": "rfdvj4ojtltljw3",
                "typeId": "commerce_assets.carousel",
                "data": {
                    "title": "Topseller",
                    "category": "topseller"
                },
                "regions": [{
                    "id": "tiles",
                    "components": [{
                            "id": "w3reö9wsjf3",
                            "typeId": "commerce_assets.productTile",
                            "data": {
                                "product": "iphone6"
                            }
                        },
                        {
                            "id": "cgn8d4ngdl4n",
                            "typeId": "commerce_assets.productTile",
                            "data": {
                                "product": "iphone7"
                            }
                        }
                    ]
                }]
            }]
        }
    ]
}]
```
  
```
import {Page} from `commerce-sdk-react/components`
import {Banner, Carousel, ProductTile} from '../../experience-components'

const PageViewer = () => {
    const page = usePage('homepage-example')
    const componentMap = {
        ['commerce_assets.banner']: Banner,
        ['commerce_assets.carousel']: Carousel,
        ['commerce_assets.productTile']: ProductTile
    }

    return <Page page={page} componentMap={componentMap} />
}
```
## Notes

1. These components are purely structural and do not add anything visually to the browser. So any additions of class names to these component templates are simply for ease of identification when viewing source and customization purposes.
2. This is a first iteration, as such the solution isn't perfect. This means that to use a page component, you have to provide the Page with a key value map of all the components that will be needed to fully render the page. In the future it would be nice to have a default implementation that automatically can load the components it needs from the components `typeId`.

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# How to Test-Drive This PR

```
> cd pages/test-commerce-sdk-react
> npm start
// Navigate to the following url --> http://localhost:3000/experience
```

On the above page you should see a rendering of the page retrieved by the API. Please note that the component implementations are samples at the most (They only render the component type id value). The idea is to see that a page and its regions/components are all being rendered. 

# Checklists




## General


## TODO:

 
## Localization
